cdk_eoapi/vpc.py

```diff
@@ -0,0 +1,50 @@
+from aws_cdk import Stack, aws_ec2
+from constructs import Construct
+
+
+class VpcStack(Stack):
+    def __init__(
+        self, scope: Construct, id: str, nat_gateway_count: int, **kwargs
+    ) -> None:
+        super().__init__(scope, id, **kwargs)
+
+        self.vpc = aws_ec2.Vpc(
+            self,
+            "vpc",
+            subnet_configuration=[
+                aws_ec2.SubnetConfiguration(
+                    name="ingress", subnet_type=aws_ec2.SubnetType.PUBLIC, cidr_mask=24
+                ),
+                aws_ec2.SubnetConfiguration(
+                    name="application",
+                    subnet_type=aws_ec2.SubnetType.PRIVATE_WITH_EGRESS,
+                    cidr_mask=24,
+                ),
+                aws_ec2.SubnetConfiguration(
+                    name="rds",
+                    subnet_type=aws_ec2.SubnetType.PRIVATE_ISOLATED,
+                    cidr_mask=24,
+                ),
+            ],
+            nat_gateways=nat_gateway_count,
```

Comment:
Note: if there is no `nat_gateways` the raster API won't be able to access any dataset!

Reply:
@vincentsarago here is my understanding

> Note: if there is no nat_gateways the raster API won't be able to access any dataset!

- The raster API needs a NAT gateway to access the outside world _if_ it's in a private subnet. 
- Whether it is in a private or public subnet is [is determined by the config parameter `public_db_subnet`. ](https://github.com/developmentseed/cdk-eoapi/blob/41120b38db37fc4701f6fdfbae48eeb425eec6d4/cdk_eoapi/pgStacInfra.py#L79)
- If `public_db_subnet` is set to `True`, then the database can reach and can be reached from the internet, along with the STAC API and the raster API, and no NAT gateway is needed at all, so there is a case for `nat_gateway=0`.
- However if `public_db_subnet` is set to `False`, we _need_ at least one NAT gateway and we must have `nat_gateway>0`.

Then if my understanding is correct, I should simply add a config validator : 

```
if not public_db_subnet and nat_gateway_count <= 0:
   raise ValueError("if the database and its associated services instances are to be located in the private subnet of the VPC, NAT gateways are needed to allow egress from the services and therefore `nat_gateway_count` has to be > 0")
```

Reply:
@sharkinsspatial 

> 
@vincentsarago Configuring subnet_type as aws_ec2.SubnetType.PRIVATE_WITH_EGRESS and passing nat_gateways > 0 to the Vpc creator should automagically create the NAT gateway https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2.SubnetType.html#private_with_egress. @emileten Can we also verify this in our deployment testing.

What you're describing is the current default configuration and yes, a NAT gets created (attached to the public subnet of the VPC) :

- [we have one NAT gateway by default](https://github.com/developmentseed/cdk-eoapi/blob/41120b38db37fc4701f6fdfbae48eeb425eec6d4/config.py#L41)
- [DB, STAC api lambda, titiler api lambda are in a private subnet by default](https://github.com/developmentseed/cdk-eoapi/blob/41120b38db37fc4701f6fdfbae48eeb425eec6d4/config.py#L38)

Reply:
@vincentsarago @sharkinsspatial lmk if you feel the above validator is enough to resolve this conversation.